---
title: "Task3_scRNA_Analysis"
author: "Tram Nguyen"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Task3_scRNA_Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  tidy.opts = list(width.cutoff = 100), tidy = TRUE
)

```

```{r load package}
library(CCBcodingTasks)
```

```{r install packages, include = FALSE}
#BiocManager::install('scRNAseq')
#BiocManager::install("SingleCellExperiment")
#BiocManager::install("OSCA.intro")
#BiocManager::install("scater")
#BiocManager::install("scran")
#BiocManager::install("bluster", force=T)
#BiocManager::install("metapod")
#BiocManager::install("devtools")
#library(devtools)
#install_github("LTLA/metapod")
#install_github("MarioniLab/scran")
```

## Analysis of single-cell RNAseq data following the “Orchestrating single-cell analysis with Bioconductor” workflow.

This vignette includes code for the third task to analyze scRNA-seq data following the [OSCA "Quick Start" workflow](https://bioconductor.org/books/3.13/OSCA.intro/analysis-overview.html#quick-start-simple) using the data from [Zeisel et al. (2015)](https://pubmed.ncbi.nlm.nih.gov/25700174/) which contains a dataset of 3005 brain cells from mice. These data include several cell types including oligodendrocytes, microglia, and neurons. Each cell was isolated using the Fluidigm C1 microfluidics system and each molecule was labeled with a unique molecular identifier (UMI). Expression was quantified by counting the number of UMI mapped to each gene. 

The workflow includes steps for QC, normalization, feature selection, dimensionality reduction, clustering, and marker gene detection. For each step, graphical summaries and brief interpretations are also included.

### 1. Load data and familiarize ourselves the SingleCellExperiment object.

```{r load data}
# Load data set
suppressMessages(library(scRNAseq))
sce.zeisel <- ZeiselBrainData()

# Following more detailed loading instructions from the OSCA book on this specific dataset (http://bioconductor.org/books/3.17/OSCA.workflows/zeisel-mouse-brain-strt-seq.html#data-loading-1), we need to merge some redundant rows. 
suppressMessages(library(scater))
sce.zeisel <- aggregateAcrossFeatures(sce.zeisel, 
                                      id=sub("_loc[0-9]+$", "", rownames(sce.zeisel)))

# Take a look at our data summary
# We have ~19K genes across 3005 cells.
sce.zeisel 
```

We can take a more detailed look at the different components of our SingleCellExperiment object. We can start with rowData, which contains a data frame of the genes and their features, such as transcript lengths or gene symbol.

```{r rowData}
head(rowData(sce.zeisel))
# Looks like we only have one column showing the featureType of the genes in this dataset.
```

Next, we can look at the metadata for our experiment, such as batch, treatment condition, type of tissue, etc.

```{r colData}
# List the different attributes of this experiment
names(colData(sce.zeisel))
```

Finally, let's take a look at the actual expression count matrix from this experiment.

In these assay matrices, cells are represented as columns, whereas in the cell metadata, cells are shown as rows. 

```{r assay}
dim(assay(sce.zeisel))
```

<br> 

### 2. Quality Control

Removing low-quality libraries is crucial for scRNA-seq data because these may contribute to spurious clustering and expression levels that could be misinterpreted as biological when they may just be technical artifacts. Low-quality cells can arise from sources such as cell damage, failure during library preparation, etc. We can use signatures in our data to try and remove this noise by filtering on quality control statistics like total counts in a cell, removing cells with few expressed genes, or high mitochondrial percentages within cells.   


```{r QC}
# Quality control (using mitochondrial genes).
library(scater)

# Check which genes are mitochondrial
is.mito <- grepl("mito", rowData(sce.zeisel)$featureType)

# For each cell, calculate QC metrics and also calculates stats specifically for the subset of mt genes too
qcstats <- perCellQCMetrics(sce.zeisel, subsets=list(Mito=is.mito))

head(qcstats)

# The sum column contains the total count for each cell and the detected column contains the number of detected genes. The subsets_Mito_percent column contains the percentage of reads mapped to mitochondrial transcripts. Finally, the altexps_ERCC_percent column contains the percentage of reads mapped to ERCC transcripts.


# View distributions of our total counts, the ones that actually passed threshold, and our mito percentages
summary(qcstats$sum)
summary(qcstats$detected)
summary(qcstats$subsets_Mito_percent)

# Identify the low quality cells based on our previous QC statistics. Filter based on Mito percent and deviations from expected ERCC spike-in controls expression levels.
filtered <- quickPerCellQC(qcstats, percent_subsets=c("altexps_ERCC_percent", 
    "subsets_Mito_percent"))

# Retain only passing cells
sce.unfiltered <- sce.zeisel
sce.pass <- sce.zeisel[,!filtered$discard]
dim(sce.pass) # 2816 passing cells

# We can also pull out the reasons why certain genes did not pass 
reasons <- perCellQCFilters(qcstats, 
                            sub.fields=c("subsets_Mito_percent", "altexps_ERCC_percent"))
colSums(as.matrix(reasons))
summary(reasons$discard)
```

Now we can create a couple of plots to look at our QC metrics.

```{r QC mito plots, fig.height = 4, fig.width = 7}
# First get our unfiltered dataset to compare to and add on filtering stats.
colData(sce.unfiltered) <- cbind(colData(sce.unfiltered), qcstats)
sce.unfiltered$discard <- filtered$discard

# View mitochondrial percentage.
plotColData(sce.unfiltered, y="subsets_Mito_percent",
        colour_by="discard") + ggtitle("Mito percent")
```

***Interpretation:*** Each point here represents a cell and is colored by whether it was discarded after our quality filtering or not. We can see that the majority of the cells that were removed (orange) had very high mitochondrial percentages and fall into the tail of the distribution. Elevated mitochondrial gene expression in a cell can be indicative of poor cell quality, stress, or technical issues during library preparation and a standard threshold is to remove cells with >= 20% mitochondrial content. We can see that there are some cells with low mito content that were also removed, which must be based on other criteria. Let's investigate another QC metric like ERCC expression.

      
```{r QC mito plots2, fig.height = 4, fig.width = 7}
# plot across tissue types now
plotColData(sce.unfiltered, x="altexps_ERCC_percent", y="subsets_Mito_percent",
        colour_by="discard", other_fields="tissue") + facet_wrap(~tissue)
```
***Interpretation:*** Now we can definitely see that high mitochondrial cells are removed, but also cells with elevated expression in the External RNA Controls Consortium (ERCC) genes are removed as well. This is because these ERCC transcripts are spiked in at known concentrations and can be used as a control. Too high of expression could indicate technical artifacts. 

<br>

### 3. Normalization

We can normalize on library size to ensure that we are comparing expression on the same scale across cells. Let's take a look at how our library sizes vary across our cells before any kind of normalization. Below, I will just show the first 100 cells in our data set.


```{R lib, fig.height = 4, fig.width = 7}
# Library size across cells before normalization
tot <- as.data.frame(cbind(colnames(assay(sce.unfiltered)), as.integer(colSums(assay(sce.unfiltered)))))

tot <- tot[1:100,]

# Plot library for first 100 cells
p1 <- ggplot(data=tot, aes(x=V1, y=as.numeric(V2))) +
  geom_point() +theme_minimal() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + xlab("Cell") + ylab("Library Size") 

print(p1)
```
***Interpretation:*** Here I am showing library sizes for the first 100 cells in our data. We can see that library size is very variable, and so we want to normalize expression counts so we can conduct comparable expression analyses. 

The first thing we can do is calculate the library size factors for our data.

```{r lib scale, fig.height = 4, fig.width = 7}
library(scater)
lib.sf.zeisel <- librarySizeFactors(sce.pass)
summary(lib.sf.zeisel)

hist(log10(lib.sf.zeisel), xlab="Log10[Size factor]", col='grey80', main= "Histogram of library size factors") 
```
***Interpretation:***  These are the “library size factor” for each cell. We can see there is variability as expected, but we also want to see that the scores are centered around zero and "balanced". This is because the use of library size factors assumes that there is no “imbalance” in the differentially expressed (DE) genes between any pair of cells. Thus, it seems like library size normalization may be sufficient for this data set. However, let's keep exploring.

Because single-cell data often has low or zero counts, we should also cluster cells with similar expression profiles and normalize within this pool to increase the size of the counts for more accurate factor estimation. We first cluster our cells, calculate the deconvolution factors, and then normalize based on these. 


```{r noramlize, fig.height = 4, fig.width = 7}
library(scran)
set.seed(100)

# Cluster similar cells based on their expression profiles, using either log-expression values or ranks.
clusters <- quickCluster(sce.pass)

# Get scaled factors and log the normalized counts based on our clustering prior
sce.zeisel <- computeSumFactors(sce.pass, cluster=clusters) 
sce.zeisel <- logNormCounts(sce.zeisel)
summary(sizeFactors(sce.zeisel))

plot(librarySizeFactors(sce.zeisel), sizeFactors(sce.zeisel), pch=16,
    xlab="Library size factors", ylab="Deconvolution factors",
    log="xy")
```

***Interpretation:*** Here we can see the positive relationship between the library size factors and the deconvolution size factors. Because composition biases can compromise DE analyses by systematically shifting the log-fold changes in one direction or another, this plot can reveal cell composition biases that stray from the correlation between the two factors. In our data, we see most cells (there are a couple deviations) have a tight correlation between deconvolution and library size factors. 

<br> 

### 4. Feature selection

We want to select highly variable genes (HVGs) that provide useful information about the biology and heterogenity of the cells, while reducing the random background noise. 

```{r HVGs, fig.height = 4, fig.width = 5}
# Model the variance of the log-expression profiles for each gene. Separate technical and biological components based on a mean-variance trend.
dec <- modelGeneVar(sce.zeisel)

# Grab the top 10% of genes with the largest biological components based on this mean-variance relationship
hvg <- getTopHVGs(dec, prop=0.1)
head(hvg)
length(hvg) #946 genes chosen

# Plot
dec.df <- dec
dec.df$Label <- "Other genes"
dec.df$Label[rownames(dec.df) %in% hvg] <- "HGVs"

dec.df <- as.data.frame(dec.df)

suppressMessages(library(dplyr))

# Plot
dec.df %>% ggplot(aes(bio, fill=Label)) +
    geom_histogram(aes(y=..count../sum(..count..)),
                 alpha=0.8,position='identity',binwidth=0.5) + theme_minimal()+
  xlab("Biological component of the variance") + ylab("Proportion of genes") 
```

***Interpretation:*** This histogram shows the distribution of the biological components of the variance for each gene calculated with the modelGeneVar function. We can see that the top 10% genes that were selected as our HGVs have the largest biological components based on this mean-variance relationship model and thus, may be the most interesting to explore in our downstream analyses.

<br>

### 5. Dimensionality Reduction with PCA

Next we want to perform dimensionality reduction across our genes to reduce the complexity of our high-dimensional scRNA-seq data for visualization and exploration. Because different genes are correlated, we can try to reduce the number of separate dimensions using something like a Principal Components Analysis (PCA).

```{r, fig.height = 4, fig.width = 7}
# Performing PCA only on the chosen HVGs.
library(scater)
set.seed(1234)

# Run a PCA and specify the genes to use for downstream functions via an extra argument subset.row=
sce.pca <- runPCA(sce.zeisel, ncomponents=25, subset_row=hvg)
reducedDimNames(sce.pca)
ncol(reducedDim(sce.pca, "PCA")) # reduced down to 25 PCs

# Plot
p.pca <- plotReducedDim(sce.pca, dimred="PCA", colour_by="level1class")
print(p.pca)
```
***Interpretation:*** Here, I am showing a PCA with the two PCs which capture the largest (PC = 35%) and second largest (PC2 = 8%) amount of variation within the dataset. Each point represents a cell that is colored based on its cell type. We can see that generally, cells are clustering with similar cell types, as we might expect. In the future, we may want to plot the top 5 PCs to explore nore dimensions of the data.

<br>

### 6. Clustering

Clustering is a technique used to group cells together based on their gene expression profiles and can help identify cellular heterogeneity.

```{r clust, fig.height = 4, fig.width = 7}
# Use the reduced dimensionality from our previous step and do clustering of cells.
library(bluster)
set.seed(1234)
sce.zeisel <- runTSNE(sce.pca, dimred="PCA")
ncol(reducedDim(sce.zeisel, "PCA"))

snn.gr <- buildSNNGraph(sce.zeisel, use.dimred="PCA")
colLabels(sce.zeisel) <- factor(igraph::cluster_walktrap(snn.gr)$membership)

# View the number of clusters and how many cells are within each.
table(colLabels(sce.zeisel))

# Plot
tsne1 <- plotTSNE(sce.zeisel, colour_by="level1class", text_by="label")
tsne2 <-plotTSNE(sce.zeisel, colour_by="label", text_by="level1class")

tsne1
tsne2

```
***Interpretation:*** Using our reduced dimensionlity values taken from the top 50 PCs from our PCA before, we can cluster our cells into expression profile groups. Here, I am showing two t-SNE plots that are just colored by cell type (top plot) and colored by the 21 clusters identified from k-means clustering (bottom plot). Interestingly, it looks like there are some cell types like oligodendrocytes that may exhibit heterogeneity in gene expression. For the most part, we see that cell types share similar expression profiles as expected.
   
<br> 

### 7. Visualization.
In the simple workflow presented in the OSCA book, they plot a UMAP to visualize the clusters. We can do the same.

```{r UMAP, fig.height = 4, fig.width = 7}
set.seed(1234)
sce <- runUMAP(sce.zeisel, dimred = 'PCA')
plotUMAP(sce, colour_by="label", text_by="level1class")
```
***Interpretation:*** Here, we are seeing similar results to our t-SNE plots. There are many difference between the two approaches but in general, UMAP preserves both local and global structures more. It often maintains more of the global structure than t-SNE plots while still providing good local clustering.

<br>

### 8. Marker gene detection
Up until now, we have investigated the marker genes that seem to be the most biologically meaningful and we labeled the clusters based on cell types so we can somewhat treat the clustering as a proxy for cell type identity. However, we may also want to dig deeper and see if there are more nuanced differences in expression between pairs of clusters. For example, we may want to know whether there are particular genes that drive the unique expression of a cluster.

```{r heatmap, fig.height = 4, fig.width = 7}
# Let's look at all the upregulated genes using pairwise wilcoxon rank sum tests.
markers <- findMarkers(sce, test.type="wilcox", direction="up", lfc=1)

# Look genes that appear more differentially expressed in cluster 4 compared to others
marker.set <- markers[["4"]]
head(marker.set[,1:8], 10) 

# Cnp, Enpp2, Klhl2 seem to the top DE expressed in cluster 4. 
top.markers <- rownames(marker.set)[marker.set$Top <= 10]

# Create a heatmap 
plotHeatmap(sce.zeisel, features=top.markers, order_columns_by="label")

```
***Interpretation:*** The heatmap of the log-expression of the top markers for cluster 4 compared to each other cluster. It is visually clear now that these genes are more DE in cluster 4 compared to the other clusters and likely contributes to producing the distinct expression profile for this group. The top three genes that were differentially expressed in cluster 4 compared to everything else were: Cnp, Enpp2, Klhl2 


Another way to look at this: we can take the top four DE genes in cluster 4 and plot their expression across all clusters.

```{r, fig.height = 4, fig.width = 7}
plotExpression(sce.zeisel, features=c("Cnp", "Enpp2",
    "Klhl2", "Mag"), x="label", colour_by="label")
```
***Interpretation:*** This plot shows the log expression of the four genes across all clusters. We can clearly see that cluster 4 has elevated expression at these genes compared to many other clusters. However, we also note that some clusters have similar expression for some of these genes (like Enpp2). Therefore, the unique clustering of cluster 4 is likely driven by the combination of many other genes too. We could continue to explore in the future. 


```{r}
sessionInfo()
```
